src/garage/torch/algos/vpg.py

```diff
@@ -135,85 +141,143 @@ def train_once(self, itr, paths):
+            tabular.record('/KLBefore', kl_before.item())
+            tabular.record('/KL', kl_after.item())
+            tabular.record('/Entropy', policy_entropy.mean().item())
+
+        undiscounted_returns = log_performance(
+            itr,
+            TrajectoryBatch.from_trajectory_list(self.env_spec, paths),
+            discount=self.discount)
+        return np.mean(undiscounted_returns)
+
+    def _train(self, itr, obs, actions, rewards, baselines, advantages):
+        """Train the algorithm with minibatch.
 
-        self.baseline.fit(paths)
-        return np.mean(average_returns)
+        Args:
+            itr (int): Iteration number.
+            obs (torch.Tensor): Observation from the environment.
+            actions (torch.Tensor): Predicted action.
+            rewards (torch.Tensor): Feedback from the environment.
+            baselines (torch.Tensor): Value function estimation at each step.
+            advantages (torch.Tensor): Expected rewards over the actions.
+
+        Returns:
+            torch.Tensor: Calculated mean value of loss
+
+        """
+        loss = self._compute_loss(itr, obs, actions, rewards, baselines,
+                                  advantages)
+
+        self._optimizer.zero_grad()
+        loss.backward()
+
+        self._optimize(itr, obs, actions, rewards, baselines, advantages)
 
-    def _compute_loss(self, itr, obs, actions, rewards, valids, baselines):
+        return loss
+
+    def _compute_loss(self, itr, obs, actions, rewards, baselines, advantages):
         """Compute mean value of loss.
 
         Args:
             itr (int): Iteration number.
             obs (torch.Tensor): Observation from the environment.
             actions (torch.Tensor): Predicted action.
             rewards (torch.Tensor): Feedback from the environment.
-            valids (list[int]): Array of length of the valid values.
             baselines (torch.Tensor): Value function estimation at each step.
+            advantages (torch.Tensor): Expected rewards over the actions.
 
         Returns:
             torch.Tensor: Calculated mean value of loss
 
         """
-        del itr
+        del itr, baselines
+        objective = self._compute_objective(advantages, obs, actions, rewards)
 
-        policy_entropies = self._compute_policy_entropy(obs)
+        return -objective.mean()
 
-        if self._maximum_entropy:
-            rewards += self._policy_ent_coeff * policy_entropies
+    def _compute_advantage(self, itr, obs, actions, rewards, valids,
+                           baselines):
+        """Compute mean value of loss.
+
+        Args:
+            itr (int): Iteration number.
+            obs (torch.Tensor): Observation from the environment.
+            actions (torch.Tensor): Predicted action.
+            rewards (torch.Tensor): Feedback from the environment.
+            valids (list[int]): Array of length of the valid values.
```

Comment:
please provide a more descriptive explanation of what valids, and rewards are.

Reply:
technically no documentation is required here because it's private

Reply:
documentation isn't required, but if it's there it should be correct